friends/src/components/SharedComponents/Form.js

```diff
@@ -0,0 +1,148 @@
+  }
+
+  addRecord = e => {
+    // prevent default
+    e.preventDefault()
+
+    // gather form data
+    let newRecord = {
+      name: this.state.name,
+      age: this.state.age,
+      email: this.state.email
+    }
+
+    // send new record to api
+    axios.post('http://localhost:5000/friends', newRecord)
+      .then(response => {
+        this.props.updateFriends(response.data)
+        this.props.history.push('/')
+      })
+      .catch(err => console.log(err))
+
+    console.log(`Form submitted data sent: ${JSON.stringify(newRecord)}`)
+    
+    // reset form fields
+    this.setState({
+      id: '',
+      name: '',
+      age: '',
+      email: ''
+    })
+  }
+
+  prePopulateForm = id => {
+    if(id === this.props.id) {
+      this.setState((
+        this.props.friend
+      ))
+    }
+  }
+
+  updateRecord = e => {
+    // prevent default
+    e.preventDefault()
+    // send updated record to api
+    axios.put(`http://localhost:5000/friends/${this.state.id}`, this.state)
+      .then(response => {
+        this.props.updateFriends(response.data)
+        this.props.history.push('/')
+      })
+      .catch(err => console.log(err))
+
+    console.log(`Form submitted data sent: ${JSON.stringify(this.state)}`)
+    
+    // reset form fields
+    this.setState({
+      id: '',
+      name: '',
+      age: '',
+      email: ''
+    })
+  }
+
+  deleteRecord = e => {
+    // prevent default
+    e.preventDefault()
+    // invoke the deleteFriend method and pass id
+    this.props.deleteFriend(this.state.id)
+    // reset form field
+    this.setState({id: ''})
+  }
+
+   submitHandler = e => {
+      e.preventDefault()
+      if(this.props.update) {
+        this.updateRecord(e)
+      } else if(this.props.delete) {
+        this.deleteRecord(e)
+      } else {
+        this.addRecord(e)
+      }
+  }
+
+  componentDidMount() {
+    if(this.props.update) {
+      this.prePopulateForm(this.props.id)
+    }
+  }
+
+  render() {
+    return (
+      <FormContainer {...this.props}>
+        <form onSubmit={this.submitHandler}>
+          {(this.props.update || this.props.delete) && 
+            <input name="id" type="number" 
+              placeholder="ID" onChange={this.inputChangeHandler} 
+              value={this.state.id} 
+            />
+          }
+          {!this.props.delete && (
+            <>
+              <input name="name" type="text" 
+                placeholder="Name" onChange={this.inputChangeHandler}
+                value={this.state.name} 
+              />
+              <input name="age" type="number" 
+                placeholder="Age" onChange={this.inputChangeHandler}
+                value={this.state.age} 
+              />
+              <input name="email" type="email" 
+                placeholder="Email" onChange={this.inputChangeHandler}
+                value={this.state.email}
+              />
+            </>
+          )}
+          <Button type="submit" {...this.props}>
```

Comment:
The button with the dynamic functionality and verb is very much in the spirit of React 🎉